modules/ppcp-button/src/Endpoint/ApproveOrderEndpoint.php

```diff
@@ -271,10 +268,79 @@ public function handle_request(): bool {
 	 *
 	 * @return void
 	 */
-	protected function toggle_final_review_enabled_setting(): void {
+	protected function toggle_final_review_enabled_setting() : void {
 		// TODO new-ux: This flag must also be updated in the new settings.
 		$final_review_enabled_setting = $this->settings->has( 'blocks_final_review_enabled' ) && $this->settings->get( 'blocks_final_review_enabled' );
 		$this->settings->set( 'blocks_final_review_enabled', ! $final_review_enabled_setting );
 		$this->settings->persist();
 	}
+
+	/**
+	 * Performs a 3DS check to verify the payment is not rejected from PayPal side.
+	 *
+	 * This method only checks, if the payment was rejected:
+	 *
+	 * - No 3DS details are present: The payment can proceed.
+	 * - 3DS details present but no rejected: Payment can proceed.
+	 * - 3DS details with a clear rejected: Payment fails.
+	 *
+	 * @param Order $order The PayPal order to inspect.
+	 * @throws RuntimeException When the 3DS check was rejected.
+	 */
+	protected function verify_three_d_secure( Order $order ) : void {
+		$payment_source = $order->payment_source();
+
+		if ( ! $payment_source ) {
+			// Missing 3DS details.
+			return;
+		}
+
+		$proceed      = ThreeDSecure::NO_DECISION;
+		$order_status = $order->status();
+		$source_name  = $payment_source->name();
+
+		/**
+		 * For GooglePay (and possibly other payment sources) we check the order
+		 * status, as it will clearly indicate if verification is needed.
+		 *
+		 * Note: PayPal is currently investigating this case.
+		 * Maybe the order status is wrong and should be ACCEPTED, in that case,
+		 * we could drop the condition and always run proceed_with_order().
+		 */
+		if ( $order_status->is( OrderStatus::PAYER_ACTION_REQUIRED ) ) {
+			$proceed = $this->threed_secure->proceed_with_order( $order );
+		} elseif ( 'card' === $source_name ) {
+			// For credit cards, we also check the 3DS response.
+			$proceed = $this->threed_secure->proceed_with_order( $order );
+		}
```

Comment:
Why instead of the `if/else` we can't just use

```
if (
    $order_status->is( OrderStatus::PAYER_ACTION_REQUIRED )
    || 'card' === $source_name
) {
    $proceed = $this->threed_secure->proceed_with_order( $order );
}
```